### ISSUE TITLE
[Bug] Catastrophic gradient explosion (NaN) in RLHF with Qwen3.5 due to 3D position_ids forcing SDPA Math fallback and BF16 collapse

### ISSUE THREAD

**Issue by ouroborosscr** [QwenLM/Qwen3.5#107]
### Description

Following up on the architectural constraints discussed in [[[QwenLM/Qwen3.5#104](https://github.com/QwenLM/Qwen3.5/issues/104)](https://github.com/QwenLM/Qwen3.5/issues/104)](https://github.com/QwenLM/Qwen3.5/issues/104), we have uncovered a severe downstream consequence when training Qwen3.5 models with RLHF (DAPO/GRPO/DPO) on long contexts (e.g., 8K - 100K tokens).

When using the default `sdpa` attention implementation with Left Padding, the model experiences a catastrophic gradient explosion, with bottom-layer gradients snowballing to $10^{28}$ or `NaN` within a few steps.

**We have fully isolated the issue and provided a reproducible repository here:** 👉 [https://github.com/ouroborosscr/Report-the-gradient-explosion-of-qwen3.5](https://github.com/ouroborosscr/Report-the-gradient-explosion-of-qwen3.5)

### 🔍 Root Cause Analysis: The Mathematical & Architectural Collision

The explosion is caused by a precise chain of architectural incompatibilities between Qwen3.5's positional embedding design, Hugging Face's attention mask materialization, and PyTorch's SDPA C++ dispatcher:

1. **3D `position_ids` forces explicit 4D Dense Mask generation:**
    
    Because Qwen3.5 utilizes 3D `position_ids` (mRoPE) to handle complex token causality, the standard 1D causal assumption fails. Consequently, in `modeling_qwen3_5.py`, Hugging Face is forced to construct a massive explicit 4D attention mask (`[Batch, 1, SeqLen, SeqLen]`) with `-3.4e38` for padding, while explicitly setting `is_causal=False`.
    
    *(Proof: Running `audit_qwen_hf.py` in the repo demonstrates that an 8192-length sequence allocates a ~0.125GB boolean mask and passes it to the underlying PyTorch SDPA).*
    
2. **PyTorch SDPA Dispatcher Downgrade (`attn_mask.has_value()`):**
    
    According to PyTorch's `sdp_utils.cpp`, the highly optimized FlashAttention kernel strictly rejects dense mask tensors (`if (attn_mask.has_value()) { return false; }`). The injection of the aforementioned 4D mask forces PyTorch to completely bypass fusion kernels and silently fall back to the naive `Math` backend.
    
3. **BF16 Precision Collapse (Swamping) in Math Backend:**
    
    The `Math` backend computes the Softmax iteratively in `bfloat16`. Over long sequences (e.g., 8k-100k tokens), the lack of FP32 SRAM accumulators (which native FlashAttention utilizes) causes severe truncation errors (swamping) in the denominator. This leads to massive numerical drift in the resulting log probabilities (`log_probs`).
    
4. **RLHF Exponential Amplification & Deep Network Cascade:**
    
    RLHF algorithms (like DPO/DAPO) rely on an exponential loss term `exp(beta * (log_pi - log_ref))`. The `Math` backend's precision drift is exponentiated, creating an initially massive gradient error at the top layer. As this error backpropagates through Qwen3.5's 32 layers of pre-trained weights, it undergoes a chain-rule multiplication that balloons the gradient to $10^{28}$ or `NaN` by the time it reaches the bottom `v_proj` layers.
    

### 💻 Reproducibility & Output Evidence

Using the minimal reproducible scripts in the provided repository:

**Test 1: Mask Materialization Audit (`audit_qwen_hf.py`)**

Plaintext

```
🔍 [Source Code Audit] What does Qwen3.5 pass to PyTorch SDPA?
➡️ Simulating DAPO Left Padding input, sequence length: 8192
🚨 [Intercepted] Qwen3.5 is calling PyTorch SDPA!
   Query shape: torch.Size([1, 8, 8192, 128])
   ⚠️ Passed Mask shape: torch.Size([1, 1, 8192, 8192])
   💀 Memory footprint of this Mask matrix alone: 0.125000 GB
   is_causal parameter: False
```

**Test 2: Gradient Explosion Verification (`train_dapo_3_debug.py`)**

- When running with default `sdpa` (Math fallback): The `ExplosionHunter` hook catches gradients exceeding `10000.0` and eventually hitting `NaN` at the bottom layers.
    
- When running with `--use_flash_attn` (`attn_implementation="flash_attention_2"`): The training remains mathematically stable. `Qwen3_5FlashAttention2` uses `varlen` to physically truncate padding via `cu_seqlens`, bypassing the 4D mask generation entirely and utilizing FP32 accumulators internally.
    

### 💡 Proposed Solution & Discussion

Currently, the only viable workaround for training Qwen3.5 with long contexts in RLHF is to strictly enforce `attn_implementation="flash_attention_2"`.

However, relying on the native PyTorch `sdpa` is fundamentally unsafe for this architecture under long-context scenarios with padding. We propose:

1. **Warning Mechanism**: `transformers` should emit a strong warning when left-padding is used with `sdpa` on models requiring 3D `position_ids`, alerting users to the silent fallback to the `Math` backend and the risk of BF16 overflow.
    
2. **Native Varlen Support**: Investigate adopting PyTorch's native `NestedTensors` or the newly introduced `torch.nn.attention.bias` mechanics to handle 3D `position_ids` without materializing a dense 4D mask, thereby keeping the native FlashAttention backend enabled.

### Reproduction

**We have fully isolated the issue and provided a reproducible repository here:** 👉 [https://github.com/ouroborosscr/Report-the-gradient-explosion-of-qwen3.5](https://github.com/ouroborosscr/Report-the-gradient-explosion-of-qwen3.5)

### Logs

```shell
1.Verify Dapo degradation

CUDA_LAUNCH_BLOCKING=1 CUDA_HOME=/usr/local/cuda-12.8 LD_PRELOAD=/opt/anaconda3/envs/scr_train2/lib/libstdc++.so.6 CUDA_VISIBLE_DEVICES=2 torchrun --nproc_per_node=1 audit_qwen_hf.py

Output:

============================================================
🔍 [源码级审计] Qwen3.5 (SDPA 模式) 到底传了什么给底层？
============================================================
➡️ 模拟 DAPO 训练输入，序列长度: 8192

🚨 [劫持成功] Qwen3.5 正在调用 PyTorch SDPA!
   Query 形状: torch.Size([1, 8, 8192, 128])
   ⚠️ Qwen 传来的 Mask 形状: torch.Size([1, 1, 8192, 8192])
   ⚠️ Mask 的数据类型: torch.bool
   💀 仅这个 Mask 矩阵就会占据显存: 0.125000 GB
   is_causal 参数: False


2.Verify gradient explosion


CUDA_LAUNCH_BLOCKING=1 CUDA_HOME=/usr/local/cuda-12.8 LD_PRELOAD=/opt/anaconda3/envs/scr_train2/lib/libstdc++.so.6 CUDA_VISIBLE_DEVICES=2 torchrun --nproc_per_node=1 train_dapo_3_debug.py  --use_lora --use_4bit


Output:

……
💥 [爆点定位] 梯度在经过 【model.layers.30.linear_attn.norm】 的反向计算后瞬间爆炸！
   传出梯度最大值: 26240.0
⚠️ [数值爆炸预警] 层: model.layers.27.self_attn.v_proj | 梯度 Max: 17280.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.27.self_attn.k_norm | 梯度 Max: 135168.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.27.self_attn.k_proj | 梯度 Max: 111616.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.27.self_attn.q_norm | 梯度 Max: 126976.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.27.self_attn.q_proj | 梯度 Max: 97280.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.23.self_attn | 梯度 Max: 67072.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.23.self_attn.o_proj | 梯度 Max: 67072.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.23.self_attn.v_proj | 梯度 Max: 13172736.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.23.self_attn.k_norm | 梯度 Max: 62390272.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.23.self_attn.k_proj | 梯度 Max: 45613056.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.23.self_attn.q_norm | 梯度 Max: 297795584.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.23.self_attn.q_proj | 梯度 Max: 242221056.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.19.self_attn | 梯度 Max: 132120576.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.19.self_attn.o_proj | 梯度 Max: 132120576.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.19.self_attn.v_proj | 梯度 Max: 48103633715200.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.19.self_attn.k_norm | 梯度 Max: 68444598829056.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.19.self_attn.k_proj | 梯度 Max: 54150947667968.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.19.self_attn.q_norm | 梯度 Max: 141836999983104.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.19.self_attn.q_proj | 梯度 Max: 204509162766336.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.15.self_attn | 梯度 Max: 131391639519232.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.15.self_attn.o_proj | 梯度 Max: 131391639519232.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.15.self_attn.v_proj | 梯度 Max: 1549526502191602335744.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.15.self_attn.k_norm | 梯度 Max: 1752440687002407403520.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.15.self_attn.k_proj | 梯度 Max: 2822351843277561397248.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.15.self_attn.q_norm | 梯度 Max: 2600990914393046777856.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.15.self_attn.q_proj | 梯度 Max: 5902958103587056517120.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.11.self_attn | 梯度 Max: 2951479051793528258560.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.11.self_attn.o_proj | 梯度 Max: 2951479051793528258560.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.11.self_attn.v_proj | 梯度 Max: 4584246707978673830485819392.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.11.self_attn.k_norm | 梯度 Max: 7969239002899635519663112192.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.11.self_attn.k_proj | 梯度 Max: 14932651723879899565970685952.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.11.self_attn.q_norm | 梯度 Max: 40852021296417549071671099392.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.11.self_attn.q_proj | 梯度 Max: 41470991316060239209120661504.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.7.self_attn | 梯度 Max: 52612451669628661683212779520.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.7.self_attn.o_proj | 梯度 Max: 52612451669628661683212779520.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.7.self_attn.v_proj | 梯度 Max: 7382797095729208034316799468109824.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.7.self_attn.k_norm | 梯度 Max: 29206669829258405410484041851863040.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.7.self_attn.k_proj | 梯度 Max: 50949412924372996104955495230472192.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.7.self_attn.q_norm | 梯度 Max: 55168154121932543553136523497963520.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.7.self_attn.q_proj | 梯度 Max: 113581493780449354374104607201689600.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.3.self_attn | 梯度 Max: 138893940965806639063190776806637568.00 | Dtype: torch.bfloat16
⚠️ [数值爆炸预警] 层: model.layers.3.self_attn.o_proj | 梯度 Max: 138893940965806639063190776806637568.00 | Dtype: torch.bfloat16
……


3.Fix

CUDA_LAUNCH_BLOCKING=1 CUDA_HOME=/usr/local/cuda-12.8 LD_PRELOAD=/opt/anaconda3/envs/scr_train2/lib/libstdc++.so.6 CUDA_VISIBLE_DEVICES=2 torchrun --nproc_per_node=1 train_dapo_3_debug.py  --use_lora --use_4bit  --use_flash_attn



Output:

……
[🔍 22:47:55] ✅ Step 1 梯度全部正常 (最大 =0.003845 @ base_model.model.model.layers.3.self_attn.v_proj.lora_B.default.weight)
……
```

### Environment Information

Model: Qwen3.5-9B (hybrid GatedDeltaNet + standard attention)
GPU: NVIDIA A100-SXM4-80GB
PyTorch: 2.9.0 / 2.10.0 (both affected)
Transformers: 5.3.0
flash-attn: 2.8.3
CUDA: 12.8
    
Repo with MRE: [[[ouroborosscr/Report-the-gradient-explosion-of-qwen3.5](https://github.com/ouroborosscr/Report-the-gradient-explosion-of-qwen3.5)](https://github.com/ouroborosscr/Report-the-gradient-explosion-of-qwen3.5)](https://github.com/ouroborosscr/Report-the-gradient-explosion-of-qwen3.5)

### Known Issue

- [x] The issue hasn't been already addressed in Documentation, Issues, and Discussions.

**Comment by ouroborosscr** [comment]
修复方法现在只找到了Flash Attention2，关于transformers层面的修复，毕竟官方这边的is_causal=False，所以我还没有找到很合适的方法

**Comment by ouroborosscr** [comment]
Flash Attention2需要使用我提报的这个版本（https://github.com/QwenLM/Qwen3.5/issues/104 ），不然会报越界错误

**Comment by JJJYmmm** [comment]
You are using qwen2 rather than qwen3.5 in your repo, right? https://github.com/ouroborosscr/Report-the-gradient-explosion-of-qwen3.5/blob/main/audit_qwen_hf.py#L42 

So I'm a bit confused why it's related to 3d position ids, and I also can't find where `is_causal` was set to False in `modeling_qwen2.py`

> Hugging Face is forced to construct a massive explicit 4D attention mask ([Batch, 1, SeqLen, SeqLen]) with -3.4e38 for padding, while explicitly setting is_causal=False.

**Comment by JJJYmmm** [comment]
Maybe you are using a (very) early version of transformers? Could you check `transformers>=5.0` (the newest one is 5.3.0)?

**Comment by ouroborosscr** [comment]
抱歉，我让gemini帮我生成的issue有些问题。我的环境和https://github.com/QwenLM/Qwen3.5/issues/104 完全一样，使用的是qwen3.5-9B和transformers==5.3.0。[[audit_qwen_hf.py#L42](ht](https://github.com/ouroborosscr/Report-the-gradient-explosion-of-qwen3.5/blob/main/audit_qwen_hf.py#L42) 是gemini帮我生成的测试脚本，他对qwen的版本有些幻觉，我会去修改验证脚本。但是后面的梯度爆炸脚本和修复脚本都是没有问题的，确实证实了最新版的transformers会出现梯度爆炸的情况。@JJJYmmm

**Comment by ouroborosscr** [comment]
@JJJYmmm 我修改了读取代码（https://github.com/ouroborosscr/Report-the-gradient-explosion-of-qwen3.5 ），得到的结果和原来一样is_causal=False
```
============================================================
🔍 [源码级审计] Qwen3.5 真实架构到底传了什么给底层？
============================================================
📦 正在加载本地真实模型: /date/sunchengrui/models/Qwen3.5-9B
`torch_dtype` is deprecated! Use `dtype` instead!
Loading weights: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████| 427/427 [00:07<00:00, 55.22it/s]
➡️ 模拟 DAPO 训练输入，序列长度: 8192

🚨 [劫持成功] 真实架构正在调用 PyTorch SDPA!
   Query 形状: torch.Size([1, 16, 8192, 256])
   ⚠️ 传来的 Mask 形状: torch.Size([1, 1, 8192, 8192])
   ⚠️ Mask 的数据类型: torch.bool
   💀 仅这个 Mask 矩阵就会占据显存: 0.125000 GB
   is_causal 参数: False
```

**Comment by JJJYmmm** [comment]
谢谢更新，但是is_causal应该是true，你可以再检查一下5.3.0 hf的attention interface，is_causal已经是作为module的一个attr而不是函数参数，你这里是false是因为https://github.com/ouroborosscr/Report-the-gradient-explosion-of-qwen3.5/blob/0b51ec292059dfbe3042485350e1b263eb8ea4f1/audit_qwen_hf.py#L9这里给了一个默认值.... 很典型的幻觉😂

**Comment by ouroborosscr** [comment]
@JJJYmmm 确实，我还需要研究一下导致sdpa降级，从而导致梯度爆炸的原因是否是qwen3.5模型
